src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509CertificateLoader.Android.cs

```diff
@@ -52,24 +53,43 @@ private static partial Pkcs12Return FromCertAndKey(CertAndKey certAndKey, Import
         {
             AndroidCertificatePal pal = (AndroidCertificatePal)certAndKey.Cert!;
 
-            if (certAndKey.Key != null)
+            if (certAndKey.Key is Pkcs12Key pkcs12Key)
             {
-                pal.SetPrivateKey(GetPrivateKey(certAndKey.Key));
-                certAndKey.Key.Dispose();
+                if (pkcs12Key is { Key: AsymmetricAlgorithm alg })
+                {
+                    pal.SetPrivateKey(GetPrivateKey(alg));
+                    certAndKey.Key.Dispose();
+                }
+                else
+                {
+                    Debug.Fail($"Unhandled key type '{pkcs12Key.Key?.GetType()?.FullName}'.");
+                    throw new CryptographicException();
+                }
             }
 
             return new Pkcs12Return(pal);
         }
 
-        private static partial AsymmetricAlgorithm? CreateKey(string algorithm)
+        private static partial Pkcs12Key? CreateKey(string algorithm, ReadOnlySpan<byte> pkcs8)
         {
-            return algorithm switch
+            switch (algorithm)
             {
-                Oids.Rsa or Oids.RsaPss => new RSAImplementation.RSAAndroid(),
-                Oids.EcPublicKey or Oids.EcDiffieHellman => new ECDsaImplementation.ECDsaAndroid(),
-                Oids.Dsa => new DSAImplementation.DSAAndroid(),
-                _ => null,
-            };
+                case Oids.Rsa or Oids.RsaPss:
+                    return new AsymmetricAlgorithmPkcs12PrivateKey(
+                        pkcs8,
+                        static () => new RSAImplementation.RSAAndroid());
+                case Oids.EcPublicKey or Oids.EcDiffieHellman:
```

Comment:
I know it was already like this, but this change is making me realize how often we repeat code.

It'd be nice (in a future change) to have just a central registration of "here's my RSA importer, or null if I don't support RSA".  Something like

```C#
private static partial KeyLoader GetKeyLoader();

...
private static partial KeyLoader GetKeyLoader()
{
    return new KeyLoader(
        rsa: pkcs8 => new ...,
        ecc: ...
```


but that's a very half-formed idea